typegate/src/runtimes/python_wasi/memory.ts

```diff
@@ -0,0 +1,102 @@
+// Copyright Metatype OÜ under the Elastic License 2.0 (ELv2). See LICENSE.md for usage.
+
+// no relative import as this file is used in https://github.com/metatypedev/python-wasi-reactor
+
+type Without<T, U> = { [P in Exclude<keyof T, keyof U>]?: never };
+type XOR<T, U> = T | U extends Record<string, unknown>
+  ? (Without<T, U> & U) | (Without<U, T> & T)
+  : T | U;
+
+export type RustResult<O = unknown[]> = XOR<{ data: O }, { error: string }>;
+
+const encoder = new TextEncoder();
+const decoder = new TextDecoder();
+
+export class Memory {
+  private memory: WebAssembly.Memory;
+  private allocate: CallableFunction;
+  private deallocate: CallableFunction;
+
+  constructor(exports: WebAssembly.Exports) {
+    this.memory = exports.memory as WebAssembly.Memory;
+    this.allocate = exports.allocate as CallableFunction;
+    this.deallocate = exports.deallocate as CallableFunction;
+  }
+
+  encode(...args: unknown[]): [number, number] {
+    const size = args.length;
+    const ptr = this.allocate(size * 2 * 4);
+    const view = new DataView(this.memory.buffer, ptr, size * 2 * 4);
+
+    for (let i = 0; i < size; i += 1) {
+      if (typeof args[i] == "string") {
+        const bytes = encoder.encode(args[i] as string);
+        const ptrV = this.allocate(bytes.length);
+        const viewV = new DataView(this.memory.buffer, ptrV, bytes.length);
+        for (let j = 0; j < bytes.length; j += 1) {
+          viewV.setUint8(j, bytes[j]);
+        }
+
+        view.setInt32(i * 2 * 4, ptrV, true);
+        view.setInt32(i * 2 * 4 + 4, bytes.length, true);
+      } else if (Number.isInteger(args[i])) {
+        const len = 4;
+        const ptrV = this.allocate(len);
+        const viewV = new DataView(this.memory.buffer, ptrV, len);
+        viewV.setInt32(0, args[i] as number, true);
```

Comment:
Suppose I have 2 consecutive integers in `args`.
```suggestion
        viewV.setInt32(0 /* why not i ?? */, args[i] as number, true);
```

Reply:
each args I, is allocating its own buffer and for integer we need to store the value at byte position 0